### ISSUE TITLE
Vim9: whitespace not enforced around lambda's -> token

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7503]
**Describe the bug**

In Vim9 script, whitespace is not enforced around the `->` token which maps the arguments of a lambda to their value, which might give unexpected results.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        eval {n: 12->len(), m: 34}
    EOF
    )

`E451` is raised:

    E451: Expected }: , m: 34}

**Expected behavior**

No error is raised.

**Environment**

 - Vim version: 8.2 Included patches: 1-2160
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(362)

**Additional context**

I think the issue comes from the fact that Vim wrongly parses the literal dictionary as a lambda.  I suspect that somehow, `: 12` is parsed as type, because we can specify a type for the argument of a lambda since [8.2.1956](https://github.com/vim/vim/releases/tag/v8.2.1956).  It was probably not an issue at the time, because we still needed the `#{}` syntax for a literal dictionary.  But since [8.2.2015](https://github.com/vim/vim/releases/tag/v8.2.2015), we don't use `#{}` anymore but simply `{}`.  I think the combination of these 2 patches create an issue which *maybe* could be fixed if we enforced whitespace around the `->` token of a lambda.

Right now, this code doesn't raise any error:
```vim
vim9script
var F = {arg1, arg2->arg1 - arg2}
echo F(5, 2)
```
    3

I think it should because there is no whitespace before and after `->`.

    var F = {arg1, arg2->arg1 - arg2}
                      ^--^
                       ✘

OTOH, this code should work:

    var F = {arg1, arg2 -> arg1 - arg2}
                       ^--^
                        ✔

Even if it doesn't fix the original issue, I think it would still make Vim9 script more readable.

**Comment by lacygoill** [comment]
> No error is raised.

Because `{n: 12->len(), m: 34}` is not a lambda but a dictionary whose keys are `n` and `m`, and whose values are `12->len()` and `34`.

**Comment by brammool** [comment]
Yes, there is confusion between using -> for method call and inside a lambda.
An alternative is to use => like in Javascript, but it does not use the {} around it: (a, b) => expression
Javascript actually allows several other variants, which doesn't make it simpler.

Java does something similar, but does use ->.

Both allow for a statement block instead of an expression.  That is a useful addition.  An alternative is to use an inline def/enddef, but that's a bit strange.

Using => instead of -> does avoid ambiguity.  And if we require parenthesis around the arguments we avoid some confusion.  So we can do:
```
let Ref = (arg: type): type => {
	    statement
	    return expr
	}
```
The simplest form, which only has arguments and an expression, would be: (arg) => expr

The implementation is a bit complicated: Whenever an opening parenthesis is encountered the matching closing parenthesis must be found and if => follows then it's a lambda, otherwise it's an expression in parenthesis.  We may need to disallow line breaks for this to work.

**Comment by lacygoill** [comment]
> Both allow for a statement block instead of an expression. That is a useful addition.

Yes, this would be useful. Currently, we need to invoke `execute()` inside a lambda to run Ex commands, but that's ugly.

> We may need to disallow line breaks for this to work.

If a new syntax is implemented, it would be nice for it to support line breaks.  This is sometimes useful when we have a complex expression, and we want to comment each component.  Example:
```vim
vim9script

com -nargs=1 SearchInMathZone SearchInMathZone(<q-args>)

def SearchInMathZone(pat: string)
    search(pat, '', 0, 0, {->
        # get the stack of syntax items under the cursor
        synstack('.', col('.'))
        # translate the IDs into names
        ->map({_, v -> synIDattr(v, 'name')})
        # skip the match if no syntax item under the cursor contains "texmathzone" in its name
        ->match('\ctexmathzone') == -1
        })
enddef
```
Although, maybe this restriction is also present in JavaScript, I don't know.

**Comment by lacygoill** [comment]
Although, I still think that enforcing whitespace around `->` could fix the issue.  If so, it's probably easier to implement than a new syntax.  OTOH, if the goal of the new syntax is not just to resolve the ambiguity, but also make Vim9 script more similar to other programming languages, then the change might be warranted.

**Comment by lacygoill** [comment]
> Although, I still think that enforcing whitespace around -> could fix the issue.

Well, I was wrong.  The issue persists.

**Comment by lacygoill** [comment]
> Well, I was wrong. The issue persists.

Sorry, the issue is fixed, but only in a `:def` function.  It persists at the script level.

So now, with the 8.2.2200 patch, this snippet works correctly:
```vim
vim9script
def Func()
    eval {n: 12->len(), m: 34}
enddef
defcompile
```
But this one is still not working:
```vim
vim9script
eval {n: 12->len(), m: 34}
```
    E451: Expected }: , m: 34}

**Comment by lacygoill** [comment]
> Using => instead of -> does avoid ambiguity. And if we require parenthesis around the arguments we avoid some confusion. So we can do:
>
>     let Ref = (arg: type): type => {
>               statement
>               return expr
>           }

This is interesting for another reason: it lets us specify a return type, which is not possible right now, even with [8.2.1956](https://github.com/vim/vim/releases/tag/v8.2.1956).  So, with this new syntax, if we make some mistake in our expression, type checking can catch it early and warn us at compile time.

**Comment by lacygoill** [comment]
So, a few remarks about the new syntax:

https://github.com/vim/vim/blob/38455a921395a56690790c8c1d28c1c43ca04c8a/runtime/doc/vim9.txt#L343-L363

The bad:

   - it doesn't work at the script level yet
   - type checking seems to ignore a wrong return type
   - forgetting a space between `):` and `return_type` seems to raise an unexpected error

The good:

   - it lets us declare a return type which was impossible before
   - it reduces the number of braces at the end (`})` → `)`) which helps making the code more readable
   - it makes it easier to distinguish a lambda from a dictionary, which improves readability
   - it makes it easier to find a lambda even when we use a lot of methods, which helps while refactoring code

Since the bad can be fixed and the good is here to stay, this is a positive change.

I wonder whether the old syntax will be deprecated.  I guess it will.  I think it should.

---

Regarding the *future* new syntax for a lambda whose rhs is a set of Ex commands (as opposed to an expression):

https://github.com/vim/vim/blob/38455a921395a56690790c8c1d28c1c43ca04c8a/runtime/doc/vim9.txt#L365-L370

I *think* it's worth considering [extra parentheses](https://github.com/vim/vim/commit/65c4415276394c871c7a8711c7633c19ec9235b1#r45405249) to help distinguish a dictionary expression from Ex commands.

**Comment by lacygoill** [comment]
> it reduces the number of braces at the end (}) → )) which helps making the code more readable

Compare:

    filter(tags, {_, v: dict<any> -> has_key(v, 'kind') ? v.kind != 'm' : true})
                                                                              ^^

Versus:

    filter(tags, (_, v: dict<any>) => has_key(v, 'kind') ? v.kind != 'm' : true)
                                                                               ^

**Comment by lacygoill** [comment]
> it doesn't work at the script level yet

Fixed by [8.2.2212](https://github.com/vim/vim/releases/tag/v8.2.2212).

> type checking seems to ignore a wrong return type
> forgetting a space between ): and return_type seems to raise an unexpected error

Fixed by [8.2.2209](https://github.com/vim/vim/releases/tag/v8.2.2209).

---

If the old syntax is meant to be deprecated, then this issue can be closed.

**Comment by lacygoill** [comment]
Since [8.2.2213](https://github.com/vim/vim/commit/7cfcd0c99c485a973d8cbff939531e430ad77f55), whitespace is only enforced for the new syntax, not the old one.  So this issue is no longer relevant.

**Comment by lacygoill** [comment]
Since it was ok to replace the `->` lambda syntax with `=>`, would it be ok to also:

   - replace the `->` method syntax with `.`?
   - deprecate the `.key` syntax?

Rationale:

   - `.` is a little less verbose than `->`
   - people coming from JavaScript or Python (which are often mentioned at `:h vim9`) would be more familiar with `.` than with `->`
   - deprecating `.key` would make Vim9 respect this principle given in the [zen of python](https://www.python.org/dev/peps/pep-0020/):

   > There should be one-- and preferably only one --obvious way to do it.

   - deprecating `.key` would make people write more consistent code

Regarding the last bullet point: right now, we can use `dict.key` when `key` is a literal key name, and `dict[key]` when `key` is an expression to evaluate, which is inconsistent.  With the new syntaxes, everybody would use the same syntax: the one using brackets.  So, `dict['key']` when `key` is a literal key name, and `dict[key]` when `key` is an expression to evaluate.

FWIW, when the new `[key]` syntax was introduced in Vim9:

https://github.com/vim/vim/blob/e5730bdcea0d4b574835f94b9813f80316590db9/runtime/doc/vim9.txt#L498-L500

I found it a little weird.  I think that's because I'm too used to `dict.key`.  If I was used to `dict['key']` instead, then I'm sure I would have found this new `[key]` natural.  IOW, I think `dict['key']` better fits in general.

---

In summary, the current syntaxes are:

    ┌───────────────────┬────────┐
    │ dictionary member │ method │
    ├───────────────────┼────────┤
    │ .key  OR  ['key'] │   ->   │
    └───────────────────┴────────┘

The new syntaxes would be:

    ┌───────────────────┬────────┐
    │ dictionary member │ method │
    ├───────────────────┼────────┤
    │      ['key']      │   .    │
    └───────────────────┴────────┘

This change was impossible in Vim script legacy because of the ambiguity with the concatenation operator `.`, but this ambiguity no longer exists in Vim9 where the concatenation operator is `..`.

**Comment by lacygoill** [comment]
> This change was impossible in Vim script legacy because of the ambiguity with the concatenation operator `.`, but this ambiguity no longer exists in Vim9 where the concatenation operator is `..`.

Sorry, disregard this last comment.  I wrote it when I hadn't thought about the benefits of deprecating `.key`.

**Comment by lacygoill** [comment]
> In summary, the current syntaxes are:

>     ┌───────────────────┬────────┐
>     │ dictionary member │ method │
>     ├───────────────────┼────────┤
>     │ .key  OR  ['key'] │   ->   │
>     └───────────────────┴────────┘

> The new syntaxes would be:

>     ┌───────────────────┬────────┐
>     │ dictionary member │ method │
>     ├───────────────────┼────────┤
>     │      ['key']      │   .    │
>     └───────────────────┴────────┘

Actually, the current syntaxes are:

    ┌──────────────────────────────┬────────┐
    │      dictionary member       │ method │
    ├──────────────────────────────┼────────┤
    │ [key]  OR  ['key']  OR  .key │   ->   │
    └──────────────────────────────┴────────┘

And the new syntaxes would be:

    ┌────────────────────┬────────┐
    │ dictionary member  │ method │
    ├────────────────────┼────────┤
    │ [key]  OR  ['key'] │   .    │
    └────────────────────┴────────┘

**Comment by brammool** [comment]
> Since it was ok to replace the `->` lambda syntax with `=>`, would it
> be ok to also:
> 
>    - replace the `->` method syntax with `.`?

No, I do like using -> for a method.  It stands out, is easy to
recognize and it does not mean the same as what "." does in other
languages.

>    - deprecate the `.key` syntax?

No, using ".member" is very common.  Using ['member'] is harder to read,
while using this syntax is needed if the argument is an expresion.

-- 
I'm not familiar with this proof, but I'm aware of a significant
following of toddlers who believe that peanut butter is the solution
to all of life's problems... 		-- Tim Hammerquist

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
